File: leetcode-go/Q051-Q100/Q091_Decode_Ways/Q091_Decode_Ways.go
package Q091_Decode_Ways

import "strconv"

func numDecodings(s string) int {
	if len(s) == 0 {
		return 0
	}
	dp := make([]int, len(s)+1)
	dp[0] = 1
	if s[0] != 0 {
		dp[1] = 1
	}
	for i := 2; i <= len(s); i++ {
		n1, _ := strconv.Atoi(s[i-1 : i])
		n2, _ := strconv.Atoi(s[i-2 : i])

		if n1 >= 1 && n1 <= 9 {
			dp[i] += dp[i-1]
		}
		if n2 >= 10 && n2 <= 26 {
			dp[i] += dp[i-2]
		}
	}
	return dp[len(s)]
}

File: src/dp/Q091_Decode_Ways.java
package dp;

public class Q091_Decode_Ways {

    public int numDecodings(String s) {
        int len = s.length();
        int[] dp = new int[len + 1];
        dp[0] = 1;
        dp[1] = (s.charAt(0) == '0') ? 0 : 1;

        for (int i = 1; i < len; i++) {
            char pCh = s.charAt(i - 1);
            char cCh = s.charAt(i);
            if (cCh != '0') dp[i + 1] += dp[i];
            if (pCh == '1' || (pCh == '2' && cCh <= '6')) dp[i + 1] += dp[i - 1];
        }

        return dp[len];
    }

}
